test/e2e/conftest.py

```diff
@@ -213,78 +212,101 @@ def worker(
+        worker: DeadlineWorker
+        if os.environ.get("USE_DOCKER_WORKER", "").lower() == "true":
+            LOG.info("Creating Docker worker")
+            worker = DockerContainerWorker(
+                configuration=worker_config,
+            )
+        else:
+            LOG.info("Creating EC2 worker")
+            ami_id = os.getenv("AMI_ID")
+            subnet_id = os.getenv("SUBNET_ID")
+            security_group_id = os.getenv("SECURITY_GROUP_ID")
+            instance_type = os.getenv("WORKER_INSTANCE_TYPE", default="t3.micro")
+            instance_shutdown_behavior = os.getenv(
+                "WORKER_INSTANCE_SHUTDOWN_BEHAVIOR", default="stop"
+            )
 
+            assert subnet_id, "SUBNET_ID is required when deploying an EC2 worker"
+            assert security_group_id, "SECURITY_GROUP_ID is required when deploying an EC2 worker"
+
+            bootstrap_resources: BootstrapResources = request.getfixturevalue("bootstrap_resources")
+            assert (
+                bootstrap_resources.worker_instance_profile_name
+            ), "Worker instance profile is required when deploying an EC2 worker"
+
+            ec2_client = boto3.client("ec2")
+            s3_client = boto3.client("s3")
+            ssm_client = boto3.client("ssm")
+            deadline_client = boto3.client("deadline")
+
+            worker = ec2_worker_type(
+                ec2_client=ec2_client,
+                s3_client=s3_client,
+                deadline_client=deadline_client,
+                bootstrap_bucket_name=bootstrap_resources.bootstrap_bucket_name,
+                ssm_client=ssm_client,
+                override_ami_id=ami_id,
+                subnet_id=subnet_id,
+                security_group_id=security_group_id,
+                instance_profile_name=bootstrap_resources.worker_instance_profile_name,
+                configuration=worker_config,
+                instance_type=instance_type,
+                instance_shutdown_behavior=instance_shutdown_behavior,
+            )
         try:
-            worker.stop()
+            worker.start()
         except Exception as e:
-            LOG.exception(f"Error while stopping worker: {e}")
-            LOG.error(
-                "Failed to stop worker. Resources may be left over that need to be cleaned up manually."
-            )
+            LOG.exception(f"Failed to start worker: {e}")
+            LOG.info("Stopping worker because it failed to start")
+            stop_worker(request, worker)
             raise
 
+        return worker
+
+    yield generate_worker
+
+
+@pytest.fixture(scope="session")
+def session_worker(
+    request: pytest.FixtureRequest, worker: Callable[[pytest.FixtureRequest], DeadlineWorker]
+) -> Generator[DeadlineWorker, None, None]:
+    session_worker: DeadlineWorker = worker(request)
+    yield session_worker
+
+    stop_worker(request, session_worker)
+
+
+@pytest.fixture(scope="function")
+def function_worker(
```

Comment:
This could be an issue with job submissions. The test fixtures are currently assuming that there is only one worker deployed at a time. If a test suite uses both `session_worker` and `function_worker`, it's possible that there can be two workers running at the same time, so tests that rely on job submissions could fail because the job went to the wrong worker, unless each worker has its own dedicated queue.

Have we tested that this does not happen? or made sure that it does not affect our tests?

Reply:
I've test and make sure test that use mixed of session and function work at the same time. All the test are e2e and have job submission

Reply:
Discussed offline, we will defer making changes to these fixtures so that job submission tests go to the worker they are meant to go to, rather than another one running at the same time, which can happen with both the session and function workers.